presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -630,6 +631,21 @@ private void addStatsListener(QueryExecution queryExecution)
     {
         Object lock = new Object();
 
+        AtomicBoolean queued = new AtomicBoolean();
+        queryExecution.addStateChangeListener(newValue -> {
+            synchronized (lock) {
+                if (newValue == QUEUED && !queued.getAndSet(true)) {
+                    stats.queryQueued();
+                }
+            }
+        });
+        synchronized (lock) {
+            // Need to do this check in case the state changed before we added the previous state change listener
+            if (queryExecution.getState() == QUEUED && !queued.getAndSet(true)) {
```

Comment:
looks like as non-obvios logic for me ;) Have you thought about writing units tests for both if conditions?